pkg/knit/cmd/irqwatch.go

```diff
@@ -0,0 +1,239 @@
+		case t := <-ticker.C:
+			lastStats, err = ih.ReadStats()
+			if err != nil {
+				return err
+			}
+			reporter.Delta(t, prevStats, lastStats)
+			prevStats = lastStats
+		}
+
+		if done {
+			break
+		}
+		if opts.maxRuns > 0 && iterCount >= opts.maxRuns {
+			break
+		}
+		iterCount++
+	}
+
+	reporter.Summary(initTs, initStats, lastStats)
+	return nil
+}
+
+type IRQReporter interface {
+	Delta(ts time.Time, prevStats, lastStats irqs.Stats)
+	Summary(initTs time.Time, prevStats, lastStats irqs.Stats)
+}
+
+func NewIRQReporter(jsonOutput bool, verbose int, cpus cpuset.CPUSet) IRQReporter {
+	if jsonOutput {
+		return &irqReporterJSON{
+			verbose: verbose,
+			cpus:    cpus,
+		}
+	}
+	return &irqReporterText{
+		verbose: verbose,
+		cpus:    cpus,
+	}
+
+}
+
+type irqReporterText struct {
+	verbose int
+	cpus    cpuset.CPUSet
+}
+
+func (w *irqReporterText) Delta(ts time.Time, prevStats, lastStats irqs.Stats) {
+	if w.verbose < 2 {
+		return
+	}
+	delta := prevStats.Delta(lastStats)
+	cpuids := w.cpus.ToSlice()
+	for _, cpuid := range cpuids {
+		counter, ok := delta[cpuid]
+		if !ok {
+			continue
+		}
+		for irqName, val := range counter {
+			if val == 0 {
+				continue
+			}
+			fmt.Printf("%v CPU=%d IRQ=%s +%d\n", ts, cpuid, irqName, val)
+		}
+	}
+}
+
+func (w *irqReporterText) Summary(initTs time.Time, prevStats, lastStats irqs.Stats) {
+	if w.verbose < 1 {
+		return
+	}
+	timeDelta := time.Now().Sub(initTs)
+	delta := prevStats.Delta(lastStats)
+	cpuids := w.cpus.ToSlice()
+
+	fmt.Printf("\nIRQ summary on cpus %v after %v\n", w.cpus, timeDelta)
+	for _, cpuid := range cpuids {
+		counter, ok := delta[cpuid]
+		if !ok {
+			continue
+		}
+		for irqName, val := range counter {
+			if val == 0 {
+				continue
+			}
+			fmt.Printf("CPU=%d IRQ=%s +%d\n", cpuid, irqName, val)
+		}
+	}
+}
+
+type irqReporterJSON struct {
+	verbose int
+	cpus    cpuset.CPUSet
+}
+
+type irqDelta struct {
+	Timestamp time.Time  `json:"timestamp"`
+	Counters  irqs.Stats `json:"counters"`
+}
+
+func (w *irqReporterJSON) Delta(ts time.Time, prevStats, lastStats irqs.Stats) {
+	if w.verbose < 2 {
+		return
+	}
+	res := irqDelta{
+		Timestamp: ts,
+		Counters:  countersForCPUs(w.cpus, prevStats.Delta(lastStats)),
+	}
+	json.NewEncoder(os.Stdout).Encode(res)
+}
+
+type irqwatchDuration struct {
+	d time.Duration
+}
+
+func (d irqwatchDuration) MarshalJSON() (b []byte, err error) {
+	return []byte(fmt.Sprintf(`"%s"`, d.d.String())), nil
+}
+
+type irqSummary struct {
+	Elapsed  irqwatchDuration `json:"elapsed"`
+	Counters irqs.Stats       `json:"counters"`
+}
+
+func (w *irqReporterJSON) Summary(initTs time.Time, prevStats, lastStats irqs.Stats) {
+	if w.verbose < 1 {
+		return
+	}
+	res := irqSummary{
+		Elapsed: irqwatchDuration{
+			d: time.Now().Sub(initTs),
+		},
+		Counters: countersForCPUs(w.cpus, prevStats.Delta(lastStats)),
+	}
+	json.NewEncoder(os.Stdout).Encode(res)
+}
+
+func countersForCPUs(cpus cpuset.CPUSet, stats irqs.Stats) irqs.Stats {
```

Comment:
May I ask to make this function public?
I use it [here](https://github.com/openshift-kni/debug-tools/pull/26/commits/08dcb7652e8ffa2ded300233ddc4d6a55e7aa512#diff-105a28bc7bb59f3ba60810661320b99cfa11962075cc2117b2f504af0cf736ebR221)

Reply:
Sure, done